Add PostgreSQL by VSHN Backup Restore documentation (#34)
appuio/appcat-docs: docs/modules/ROOT/pages/vshn-managed/postgresql/restore.adoc

= Restore a PostgreSQL service

IMPORTANT: The restores can only be triggered from an instance in the same namespace.
If the old instance has already been deleted the backups are deleted as well.

. TODO: get the name of the backup you want to restore...

. Apply the following object on your namespace, as specified by its YAML description.
+
.Example to restore a PostgreSQL instance
[source,yaml]
----
apiVersion: vshn.appcat.vshn.io/v1
kind: VSHNPostgreSQL
metadata:
  name: pgsql-app1-restore
spec:
  parameters:
    restore:
      claimName: pgsql-app1-prod <1>
      backupName: pgsql-app1-prod-dx4fd-2023-02-17-08-50-01 <2>
    backup:
      schedule: '0 22 * * *'
    service:
      majorVersion: "15"
      pgSettings:
        timezone: Europe/Zurich
    size:
      cpu: "600m"
      memory: "3500Mi"
      disk: "80Gi"
  writeConnectionSecretToRef:
    name: postgres-creds-restored
  compositionRef:
    name: vshnpostgresrestore.vshn.appcat.vshn.io <3>
----

> <1> The claim name of the instance you want to restore from

Reviewer: Is there any reason why this is called "claim" other than because Crossplane names it like that? This isn't too user friendly IMHO, we should rather name this "instance".

= Restore a PostgreSQL service

IMPORTANT: The restores can only be triggered from an instance in the same namespace.
If the old instance has already been deleted the backups are deleted as well.

. Get the name of the backup you want to restore by listing all available backups.
+
[source,bash]
----

> kubectl -n postgres get vshnpostgresbackups

Reviewer: ```suggestion
kubectl -n <your-namespace> get vshnpostgresbackups
```

BACKUP NAME                                 DATABASE INSTANCE   STORED TIME               STATUS     AGE

pgsql-app1-prod-pk8k4-2023-03-05-13-05-00   pgsql-app1-prod     2023-03-05T13:05:15.755Z  Completed  38s
pgsql-app1-dev-pk8k4-2023-03-01-16-52-11    pgsql-app1-dev                                Failed     28s
----
NOTE: The list will show all backups from all instances in this namespace. In the example above there are two database instances with one backup each.

> 

Reviewer 2: ```suggestion
IMPORTANT: Restoring a backup will create a new instance and not revert the old one.
```

. Apply the following object on your namespace, as specified by its YAML description.
+
.Example to restore a PostgreSQL instance
[source,yaml]
----
apiVersion: vshn.appcat.vshn.io/v1
kind: VSHNPostgreSQL
metadata:
  name: pgsql-app1-restore

>   namespace: postgres

Reviewer: ```suggestion
  namespace: <your-namespace>
```

spec:
  parameters:
    restore:

>       claimName: pgsql-app1-prod <1>

Reviewer: Is there a reason why we - again - use the term "claim" here? I thought we had a discussion earlier that this term is specific to Crossplane and not for the end-user, as it's not used elsewhere and we also don't use it in other user-facing APIs afaik.

I guess this is namespaced? I mean, could it be a security issue and being able to take over other already existing instances?

Reviewer 3: > Is there a reason why we - again - use the term "claim" here? I thought we had a discussion earlier that this term is specific to Crossplane and not for the end-user, as it's not used elsewhere and we also don't use it in other user-facing APIs afaik.

I agree with you, we should not use crossplane terms where possible. Unfortunately we do haven't updated the claim yet so I will have to open a task for that. Meanwhile is it ok to leave it as is?

> I guess this is namespaced? I mean, could it be a security issue and being able to take over other already existing instances?

This is namespace scoped therefore the instances are accessed by any user having access to the namespace. Personally I don't see a problem here. If the customer wants different access to his databases then he should create them in different namespaces.

      backupName: pgsql-app1-prod-pk8k4-2023-03-01-16-52-02 <2>
    backup:
      schedule: '0 22 * * *'
    service:
      majorVersion: "15"
      pgSettings:
        timezone: Europe/Zurich
    size:
      cpu: "600m"
      memory: "3500Mi"
      disk: "80Gi"
  writeConnectionSecretToRef:
    name: postgres-creds-restored
  compositionRef:
    name: vshnpostgresrestore.vshn.appcat.vshn.io <3>
----
<1> The name of the instance you want to restore from
<2> The backup name you want to restore
<3> You need to specify that this is not a normal instance but a restored one
+

> NOTE: The restore process is the same as ordering a new instance. The only difference is the backup information in `spec.parameters.restore` object.

Reviewer: Isn't `compositionRef` also different compared to ordering a new instance?

Reviewer 3: you are right, updated the Note.

= Restore a PostgreSQL service

IMPORTANT: The restores can only be triggered from an instance in the same namespace.
If the old instance has already been deleted the backups are deleted as well.

. Get the name of the backup you want to restore by listing all available backups.
+
[source,bash]
----
kubectl -n postgres get vshnpostgresbackups

> BACKUP NAME                                 DATABASE INSTANCE     STATUS     AGE

Reviewer: I'd expect a printer column from when this backup actually is. Yes, it's in the name, but that's not very user-friendly. Could we expose `status.process.timing.stored` here?

Reviewer 3: Yes we can do that.

pgsql-app1-prod-pk8k4-2023-03-05-13-05-00   pgsql-app1-prod       Completed  38s
pgsql-app1-dev-pk8k4-2023-03-01-16-52-11    pgsql-app1-dev        Failed     28s
----
NOTE: The list will show all backups from all instances in this namespace. In the example above there are two database instances with one backup each.

> . Choose a specific backup and get its details.

Reviewer: For what do I need the details exactly?

Reviewer 3: This step is not necessary for the subsequent steps but it's a way to tell the customer he can have more details regarding its backups.

Reviewer: Then please add it to a separate section. Step-by-step tutorials should come to a point, not adding any unnecessary detours.

Reviewer 3: I added a new page as it seemed to me it's not a good idea to show how backups work in a restore page.

+
[source,bash]
----
kubectl -n postgres get vshnpostgresbackups pgsql-app1-prod-pk8k4-2023-03-05-13-05-00 -oyaml
----
+
[source,yaml]
----
apiVersion: api.appcat.vshn.io/v1
kind: VSHNPostgresBackup
metadata:
  name: pgsql-app1-prod-pk8k4-2023-03-05-13-05-00
  namespace: postgres
status:
  databaseInstance: pgsql-app1-prod
  process:
    failure: "" <1>
    jobPod: pgsql-app1-prod-pk8k4-backup-27967024-6xjpq
    status: Completed
    timing:
      end: "2023-03-05T13:05:15.61228Z"
      start: "2023-03-05T13:05:09.519369Z"
      stored: "2023-03-05T13:05:15.755Z"
----

> <1> The failure message in case the job fails.

Reviewer: ```suggestion
<1> The failure message in case the backup job failed.
```

. Apply the following object on your namespace, as specified by its YAML description.
+

> .Example to restore a PostgreSQL instance

Reviewer: As I understand, that's just yet another "normal" service ordering object, but with restore parameters in it? IMHO that would need to be explained...

[source,yaml]
----
apiVersion: vshn.appcat.vshn.io/v1
kind: VSHNPostgreSQL
metadata:
  name: pgsql-app1-restore
  namespace: postgres
spec:
  parameters:
    restore:
      claimName: pgsql-app1-prod <1>
      backupName: pgsql-app1-prod-pk8k4-2023-03-01-16-52-02 <2>
    backup:
      schedule: '0 22 * * *'
    service:
      majorVersion: "15"
      pgSettings:
        timezone: Europe/Zurich
    size:
      cpu: "600m"
      memory: "3500Mi"
      disk: "80Gi"
  writeConnectionSecretToRef:
    name: postgres-creds-restored
  compositionRef:
    name: vshnpostgresrestore.vshn.appcat.vshn.io <3>
----

> <1> The claim name of the instance you want to restore from

Author: @zugao You overwrote my change here again :)

Can you please remove the word `claim` from this line? Thanks.

<2> The backup name you want to restore

> <3> You need to specify that this is not a normal instance but a restored one

Reviewer: Can I remove that again after the restore? Maybe the same questions as the next one for line 78.

Reviewer 3: > Can I remove that again after the restore? Maybe the same questions as the next one for line 78.
By remove you mean edit the claim and erase the restore part?

Author: See above, it might be possible to switch the composition and thus then also be able to remove the restore part. But we probably need to investigate if that actually works or not...

Reviewer: We somehow have to be able to restore the original object again...

Author: With the current restore strategy that we defined that's not possible. A restore is always into a new instance. https://products.docs.vshn.ch/products/appcat/data_protection.html#_restore_process

Reviewer 3: We might want to move the restore object from `spec` to `status` after the restore process finished. I need to check  that to see if it's feasible. Will it work for you that way? Otherwise we will have to use composition functions.

Reviewer: I don't think this is what I meant. I meant:

1. Specify the resource as described here, with all the needed parameters for the restore
2. Once the restore is finished: Remove all the restore related parameters again: `spec.parameters.restore` and `spec.compositionRef`.

Oh, and `spec.writeConnectionSecretToRef` is also a problem now, because it has a different name.

Put yourself into a users' perspective: You'd have to restore from a backup and after that, you have to configure your application for a different parameter and have a "tainted" resource in your deployment pipeline. Just doesn't feel good.

Reviewer: An idea: We could change the restore process to:

1. Create a new restore instance, just like described here
2. Instruct the user to do some database dumping from the restore instance to the original instance
3. Remove the restore instance again

That could probably work as a suitable solution.

Reviewer 3: This is very confusing. I would investigate a solution with Composition Functions and remove the restore composition altogether. We already tested it and it seems to work though it's in alpha for now.

Reviewer 4: To summarize the issues here and that we're sure that we talk about the same thing.

There are two issues with this:

* The restored instance should permanently use a different `compositionRef`. We should figure out if we can change it to the normal `compositionRef` after the restore. This probably should just work.
* The documentation isn't clear enough that restoring will create a **new** instance and not revert the old one.

There are also related discussion points, but these aren't blocking but input we should look into

* Maybe we should provide a way to restore/revert the original instance
* Using different `compositionRef` is ugly, we should find a way around that. Maybe with composition functions

Did I understand both sides correctly?

Reviewer: Yes, from my side. And: When the user restores into a new instance, the old instance would have to be deleted to save costs. But with the deletion, also older backups are gone, which is an issue. We should find a way to keep older backups, maybe transfer them to the new instance? Or, as written above, find a way to restore to the original instance.

Reviewer 2: A follow-up work item APPCAT-382 as well as 2 new feature ideas in the product board are created. 
Shall we work on the documentation together in a separate session, @tobru ?

> . After the restore the new instance will behave exactly the same as any other PostgreSQL by VSHN instance.

Reviewer: That's not an actionable step, isn't it? So no need to be in the list. How can I make this object "normal" looking, without the restore stuff in it? Aka how can I make it GitOps-ready again?

Reviewer 3: > That's not an actionable step, isn't it? So no need to be in the list.
I will remove the action point.

> How can I make this object "normal" looking, without the restore stuff in it? Aka how can I make it GitOps-ready again?

If I understood you correctly, for now we cannot make it GitOps ready. This is partly due to the limitations of Crossplane. We are using a second composition for the restore process. @Kidswiss do you see it differently?

Author: It might be possible to switch the composition back to the default one after the restore has finished. But haven't tested that.

Reviewer: Please test it, sounds like a very important thing. This isn't the "original" object anymore, which would completely break CI/CD pipelines and GitOps.

Author: I'm not quite sure what you mean by "original" object.

The restore needs to be done with a new instance defined by a new object anyway. And that new object can also be managed via GitOps.
